walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,92 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.PriorityQueue;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return new ArrayList<>();
+    } 
+    
+    PriorityQueue<TimeRange> busyTimeRanges = getBusyTimeRanges(events, request);
+    ArrayList<TimeRange> combinedBusyTimeRanges = getCombinedBusyTimeRanges(busyTimeRanges);
+    Collection<TimeRange> result = getAvailableTimeRanges(combinedBusyTimeRanges, request);
+    return result;
+  }
+
+  private boolean eventHasRequestAttendees(Collection<String> requestAttendees, Event event) {
+    Set<String> eventAttendees = event.getAttendees();
+
+    for (String requestAttendee : requestAttendees) {
+      if (eventAttendees.contains(requestAttendee)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+  
+  private PriorityQueue<TimeRange> getBusyTimeRanges(Collection<Event> events, MeetingRequest request) {
+    PriorityQueue<TimeRange> busyTimeRanges = new PriorityQueue<>(TimeRange.ORDER_BY_START);
+    Collection<String> requestAttendees = request.getAttendees();
+
+    for (Event event : events) {
+      if (eventHasRequestAttendees(requestAttendees, event)) {
+        TimeRange eventTimeRange = event.getWhen();
+        busyTimeRanges.add(eventTimeRange); 
+      }
+    }
+
+    return busyTimeRanges;
+  }
+
+  private ArrayList<TimeRange> getCombinedBusyTimeRanges(PriorityQueue<TimeRange> busyTimeRanges) {
+    ArrayList<TimeRange> combinedBusyTimeRanges = new ArrayList<>();
+
+    while (!busyTimeRanges.isEmpty()) {
+      TimeRange busyTimeRange = busyTimeRanges.poll();
+      int startTime = busyTimeRange.start();
+      int endTime = busyTimeRange.end();
+      
+      while (!busyTimeRanges.isEmpty() && busyTimeRanges.peek().overlaps(busyTimeRange)) {
+        TimeRange nextOverlappingBusyTimeRange = busyTimeRanges.poll();
+        endTime = Math.max(endTime, nextOverlappingBusyTimeRange.end());
+      }
+      
+      TimeRange newTimeRange = TimeRange.fromStartEnd(startTime, endTime, false);
+      combinedBusyTimeRanges.add(newTimeRange);
+    }
+
+    return combinedBusyTimeRanges;
+  }
+
+  private void processNewTimeRange(ArrayList<TimeRange> availableTimeRanges, TimeRange newTimeRange, MeetingRequest request) {
+    if (newTimeRange.duration() >= request.getDuration()) {
+      availableTimeRanges.add(newTimeRange);
+    }
+  }
+
+  private Collection<TimeRange> getAvailableTimeRanges(ArrayList<TimeRange> combinedBusyTimeRanges, MeetingRequest request) {
+    ArrayList<TimeRange> availableTimeRanges = new ArrayList<>();
```

Comment:
okay!

Reply:
its better to change it to `return Collections.emptyList()` for easier to read that if that condition were to happen, the function will return empty list

Reply:
or any other options of your choices: https://stackoverflow.com/questions/28406921/java-return-an-empty-list

Reply:
Is `Collections.emptyList()` more applicable at line 26 where I am checking if `request.getDuration() > TimeRange.WHOLE_DAY.duration()` and if that is the case, I need to return an empty collection?

For the if block at line 83, I am checking if `combinedBusyTimeRanges.size() == 0` and if that is the case, I need to add the whole day to `availableTimeRanges` instead of empty collection. Is there a better way of returning a collection of 1 item?

Reply:
usually, it is a good practice to use immutables in a project, and it just happen that ImmutableList can return one element array in that way
```ImmutableList.of(TimeRange.WHOLE_DAY)```
however in this case, we can leave it this way for now